src/Microsoft.CodeAnalysis.Analyzers/Core/ImmutableObjectMethodAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+
+        private static readonly ImmutableArray<string> s_immutableMethodNames = ImmutableArray.Create(
+            "Add",
+            "Remove",
+            "Replace",
+            "With");
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var compilation = context.Compilation;
+                var builder = ImmutableArray.CreateBuilder<INamedTypeSymbol>();
+                var provider = WellKnownTypeProvider.GetOrCreate(compilation);
+                AddIfNotNull(builder, provider.GetOrCreateTypeByMetadataName(SolutionFullName));
+                AddIfNotNull(builder, provider.GetOrCreateTypeByMetadataName(ProjectFullName));
+                AddIfNotNull(builder, provider.GetOrCreateTypeByMetadataName(DocumentFullName));
+                AddIfNotNull(builder, provider.GetOrCreateTypeByMetadataName(SyntaxNodeFullName));
+                AddIfNotNull(builder, provider.GetOrCreateTypeByMetadataName(CompilationFullName));
+                var immutableTypeSymbols = builder.ToImmutable();
+                if (immutableTypeSymbols.Length > 0)
+                {
+                    context.RegisterOperationAction(context => AnalyzeInvocationForIgnoredReturnValue(context, immutableTypeSymbols), OperationKind.Invocation);
+                }
+            });
+
+            static void AddIfNotNull(ImmutableArray<INamedTypeSymbol>.Builder builder, INamedTypeSymbol? symbol)
+            {
+                if (symbol is not null)
+                {
+                    builder.Add(symbol);
+                }
+            }
+        }
+
+        public static void AnalyzeInvocationForIgnoredReturnValue(OperationAnalysisContext context, ImmutableArray<INamedTypeSymbol> immutableTypeSymbols)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            if (invocation.Parent is not IExpressionStatementOperation)
+            {
+                return;
+            }
+
+            // If the method doesn't start with something like "With" or "Replace", quit
+            string methodName = invocation.TargetMethod.Name;
+            if (!s_immutableMethodNames.Any(n => methodName.StartsWith(n, StringComparison.Ordinal)))
+            {
+                return;
+            }
+
+            INamedTypeSymbol? type = null;
+            if (invocation.Instance is not null)
+            {
+                type = invocation.Instance.Type as INamedTypeSymbol;
+            }
+            else if (invocation.TargetMethod.IsExtensionMethod && invocation.TargetMethod.Parameters is [{ Type: var parameterType }, ..])
+            {
+                type = parameterType as INamedTypeSymbol;
+            }
```

Comment:
@mavasani Does this logic sound correct? I'm surprised there is no property on IInvocationOperation to get this info directly (or at least I couldn't find it)

Reply:
Can you instead use https://github.com/dotnet/roslyn-analyzers/blob/b7bb138809d5a7d31508fe0cd86d59ed4c864764/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L23-L28

Reply:
It looks like `GetReceiverType` implementation is incorrect when `beforeConversion` is false.

`firstArg.Type` is always null. I'll use this extension and fix its implementation.